File: src/math/vector.c
// Log category, used to filter logs
#define LOG_CATEGORY "Math"

#include "vector.h"

#include "check.h"
#include "quat.h"
#include <math.h>

// Returns the length of the vactor
// NOTE: When comparing two vectors, vec_len_squared may be preferrable as it's
// much faster.
float vec2_len(vec2 v) {
    return sqrt(vec2_len_squared(v));
}
float vec3_len(vec3 v) {
    return sqrt(vec3_len_squared(v));
}
float vec4_len(vec4 v) {
    return sqrt(vec4_len_squared(v));
}

// Returns the squared length of the vactor
float vec2_len_squared(vec2 v) {
    return square(v.x) + square(v.y);
}
float vec3_len_squared(vec3 v) {
    return square(v.x) + square(v.y) + square(v.z);
}
float vec4_len_squared(vec4 v) {
    return square(v.x) + square(v.y) + square(v.z) + square(v.w);
}

// Normalizes the vector, making its length 1.
vec2 vec2_normalize(vec2 v) {
    float len = vec2_len(v);

    check_return(len > 0, "Attempting to normalize a vec2 of 0 length is undefined.", vec2_zero);

    return (vec2){ .x = v.x / len, .y = v.y / len };
}
vec3 vec3_normalize(vec3 v) {
    float len = vec3_len(v);

    check_return(len > 0, "Attempting to normalize a vec3 of 0 length is undefined.", vec3_zero);

    return (vec3){ .x = v.x / len, .y = v.y / len, .z = v.z / len };
}
vec4 vec4_normalize(vec4 v) {
    float len = vec4_len(v);

    check_return(len > 0, "Attempting to normalize a vec4 of 0 length is undefined.", vec4_zero);

    return (vec4){ .x = v.x / len, .y = v.y / len, .z = v.z / len, .w = v.w / len };
}

// Adds 2 vectors together
vec2 vec2_add(vec2 v1, vec2 v2) {
    return (vec2){ .x = v1.x + v2.x, .y = v1.y + v2.y };
}
vec3 vec3_add(vec3 v1, vec3 v2) {
    return (vec3){ .x = v1.x + v2.x, .y = v1.y + v2.y, .z = v1.z + v2.z };
}
vec4 vec4_add(vec4 v1, vec4 v2) {
    return (vec4){ .x = v1.x + v2.x, .y = v1.y + v2.y, .z = v1.z + v2.z, v1.w + v2.w };
}

// Subtracts v2 from v1
vec2 vec2_sub(vec2 v1, vec2 v2) {
    return (vec2) { .x = v1.x - v2.x, .y = v1.y - v2.y };
}
vec3 vec3_sub(vec3 v1, vec3 v2) {
    return (vec3) { .x = v1.x - v2.x, .y = v1.y - v2.y, .z = v1.z - v2.z };
}
vec4 vec4_sub(vec4 v1, vec4 v2) {
    return (vec4) { .x = v1.x - v2.x, .y = v1.y - v2.y, .z = v1.z - v2.z, .w = v1.w - v2.w };
}

// Multiplies v by the scalar x
vec2 vec2_mul(vec2 v, float f) {
    return (vec2) { .x = v.x * f, .y = v.y * f };
}
vec3 vec3_mul(vec3 v, float f) {
    return (vec3) { .x = v.x * f, .y = v.y * f, .z = v.z * f };
}
vec4 vec4_mul(vec4 v, float f) {
    return (vec4) { .x = v.x * f, .y = v.y * f, .z = v.z * f, .w = v.w * f };
}

// Returns the dot product between v1 and v2
float vec2_dot(vec2 v1, vec2 v2) {
    return (v1.x * v2.x) + (v1.y * v2.y);
}
float vec3_dot(vec3 v1, vec3 v2) {
    return (v1.x * v2.x) + (v1.y * v2.y) + (v1.z * v2.z);
}
float vec4_dot(vec4 v1, vec4 v2) {
    return (v1.x * v2.x) + (v1.y * v2.y) + (v1.z * v2.z) + (v1.w * v2.w);
}

// Returns the cross product of v1 and v2.
vec3 vec3_cross(vec3 v1, vec3 v2) {
    return (vec3){ .x = (v1.y * v2.z) - (v1.z * v2.y), .y = (v1.z * v2.x) - (v1.x * v2.z), .z = (v1.x * v2.y) - (v1.y * v2.x) };
}

vec2 vec2_rotate(vec2 v, float angle) {
    return (vec2) {
        .x = v.x * cos(angle) - v.y * sin(angle),
        .y = v.x * sin(angle) + v.y * cos(angle)
    };
}
vec3 vec3_rotate(vec3 v, quat q) {
    vec3 u = (vec3) { .x = q.x, .y = q.y, .z = q.z };
    float s = q.w;

    return vec3_add(
            vec3_add(
                vec3_mul(u, 2.0f * vec3_dot(u, v)),
                vec3_mul(v, square(s) - vec3_dot(u, u))
            ), vec3_mul(vec3_cross(u, v), 2 * s)
        );
}
vec4 vec4_rotate(vec4 v, quat q) {
    return (vec4) { .xyz=vec3_rotate(v.xyz, q) };
}

File: src/math/vector.js
class vec2 {
    constructor() {
        this.x = (this.arguments.length > 0) ? this.arguments[0] : 0;
        this.y = (this.arguments.length > 1) ? this.arguments[1] : 0;
    }
    len_squared() {
        return Math.pow(this.x, 2) + Math.pow(this.y, 2);
    }
    len() {
        return Math.sqrt(this.len_squared());
    }
    normalize() {
        var l = this.len()
        if(l == 0) {
            throw new UserException('Attempting to normalize a vec2 of 0 length is undefined.')
        }

        x /= l;
        y /= l;
    }
    add(vec) {
        return new vec2(x + vec.x, y + vec.y);
    }
    sub(vec) {
        return new vec2(x - vec.x, y - vec.y);
    }
    dot(vec) {
        return (x * vec.x) + (y * vec.y);
    }
}
class vec3 {
    constructor() {
        this.x = (this.arguments.length > 0) ? this.arguments[0] : 0;
        this.y = (this.arguments.length > 1) ? this.arguments[1] : 0;
        this.z = (this.arguments.length > 2) ? this.arguments[2] : 0;
    }
    len_squared() {
        return Math.pow(this.x, 2) + Math.pow(this.y, 2) + Math.pow(this.z, 2);
    }
    len() {
        return Math.sqrt(this.len_squared());
    }
    normalize() {
        var l = this.len()
        if(l == 0) {
            throw new UserException('Attempting to normalize a vec3 of 0 length is undefined.')
        }

        x /= l;
        y /= l;
        z /= l;
    }
    add(vec) {
        return new vec3(x + vec.x, y + vec.y, z + vec.z);
    }
    sub(vec) {
        return new vec3(x - vec.x, y - vec.y, z - vec.z);
    }
    dot(vec) {
        return (x * vec.x) + (y * vec.y) + (z * vec.z);
    }
    cross(vec) {
        return vec3((y * vec.z) - (z * vec.y), (z * vec.x) - (x * vec.z), (x * vec.y) - (y * vec.x))
    }
}
class vec4 {
    constructor() {
        this.x = (this.arguments.length > 0) ? this.arguments[0] : 0;
        this.y = (this.arguments.length > 1) ? this.arguments[1] : 0;
        this.z = (this.arguments.length > 2) ? this.arguments[2] : 0;
        this.w = (this.arguments.length > 3) ? this.arguments[3] : 0;
        if
    }
    len_squared() {
        return Math.pow(this.x, 2) + Math.pow(this.y, 2) + Math.pow(this.z, 2) + Math.pow(this.w, 2);
    }
    len() {
        return Math.sqrt(this.len_squared());
    }
    normalize() {
        var l = this.len()
        if(l == 0) {
            throw new UserException('Attempting to normalize a vec4 of 0 length is undefined.')
        }

        x /= l;
        y /= l;
        z /= l;
        w /= l;
    }
    add(vec) {
        return new vec4(x + vec.x, y + vec.y, z + vec.z, w + vec.w);
    }
    sub(vec) {
        return new vec4(x - vec.x, y - vec.y, z - vec.z, w - vec.w);
    }
    dot(vec) {
        return (x * vec.x) + (y * vec.y) + (z * vec.z) + (w * vec.w);
    }
}
